### PR TITLE
Circuit API

### PR DESCRIPTION
**Pull request by imeckler**

### CODE REVIEW

File: kimchi/src/prover.rs [hunk 1]
```diff
@@ -73,13 +80,23 @@ where
...
+        };
         let d1_size = index.cs.domain.d1.size as usize;
         // TODO: rng should be passed as arg
         let rng = &mut rand::rngs::OsRng;
 
         // double-check the witness
-        if cfg!(test) {
+        if true
+        /* cfg!(test) */
+        {
```
Comment (mimoo) [inline kimchi/src/prover.rs:99]: we should just replace all of these with `if cfg!(debug_assertions)`.

File: kimchi/src/prover.rs [hunk 2]
```diff
@@ -146,10 +167,33 @@ where
...
+            match &blinders[i] {
+                None => index
+                    .srs
+                    .commit_evaluations(index.cs.domain.d1, &e, None, rng),
+                Some(b) => {
+                    let c = index
+                        .srs
+                        .commit_evaluations_non_hiding(index.cs.domain.d1, &e, None);
+                    (
+                        c.zip(b).unwrap().map(|(g, b)| {
```
Comment (mimoo) [inline kimchi/src/prover.rs:180]: `c.zip(b).unwrap().map(|(g, b)|`

that's a lot of one-letter variable names :p

File: circuit-construction/src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,988 @@
...
+    }
+}
+
+pub struct ShiftedScalar<F>(Var<F>);
+
+pub struct GateSpec<F: FftField> {
+    pub typ: GateType,
+    pub row: [Var<F>; COLUMNS],
+    pub c: Vec<F>,
+}
```
Comment (mimoo) [inline circuit-construction/src/lib.rs:122]: we already have this type as `CircuitGate`

File: circuit-construction/src/lib.rs [hunk 4]
```diff
@@ -0,0 +1,988 @@
...
+        final_row[2] = states[states.len() - 1][2];
+        self.gate(GateSpec {
+            typ: kimchi::circuits::gate::GateType::Zero,
+            c: vec![],
+            row: final_row,
+        });
+
+        states.pop().unwrap()
+    }
+}
```
Comment (mimoo) [inline circuit-construction/src/lib.rs:761]: I think these gadgets should live within their respective files, not here. Poseidon, for example, already has a gadget here https://github.com/o1-labs/proof-systems/blob/master/kimchi/src/circuits/polynomials/poseidon.rs#L95

File: circuit-construction/src/lib.rs [hunk 5]
```diff
@@ -0,0 +1,988 @@
...
+        self.gates.len()
+    }
+
+    fn gate(&mut self, g: GateSpec<F>) {
+        self.gates.push(g);
+    }
+}
+
+impl<F: FftField> System<F> {
+    pub fn gates(&self) -> Vec<CircuitGate<F>> {
```
Comment (querolita) [inline circuit-construction/src/lib.rs:809]: where is the wiring handled?

**Review by querolita — COMMENTED** [review]
Awesome line of work 🦾

File: circuit-construction/src/lib.rs [hunk 6]
```diff
@@ -0,0 +1,988 @@
...
+    pub gates: Vec<GateSpec<F>>,
+}
+
+pub struct WitnessGenerator<F> {
+    pub rows: Vec<Row<F>>,
+}
+
+type Row<V> = [V; COLUMNS];
+
+pub trait Cs<F: FftField + PrimeField> {
```
[the remark this answers is not in the document]
Comment (mimoo) [reply circuit-construction/src/lib.rs:143]: +1, that's not a great name

File: circuit-construction/benches/example_proof.rs [hunk 7]
```diff
@@ -0,0 +1,128 @@
...
+    let actual = sys.scalar_mul(zero, base, scalar);
+
+    let preimage = sys.var(|| witness.as_ref().unwrap().preimage);
+    let actual_hash = sys.poseidon(constants, vec![preimage, zero, zero])[0];
+
+    sys.assert_eq(actual.0, public_input[0]);
+    sys.assert_eq(actual.1, public_input[1]);
+    sys.assert_eq(actual_hash, public_input[2]);
+
+    sys.zk()
```
Comment (o1-david) [inline circuit-construction/benches/example_proof.rs:69]: I don't think this should be part of the circuit :o and we should continue to do it transparently

File: circuit-construction/src/lib.rs (hunk 4 above)
Comment (jspada) [reply circuit-construction/src/lib.rs:761]: Agree-- it's nice to have clean separation and all of a gate's code in a single file or module.  I think the proof-system's hooks into the gate should be as clean as possible and not have to know internal details about the gate.

File: circuit-construction/src/lib.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (jspada) [reply circuit-construction/src/lib.rs:143]: Good one.

File: kimchi/src/prover.rs [hunk 8]
```diff
@@ -73,13 +80,23 @@ where
         mut witness: [Vec<ScalarField<G>>; COLUMNS],
         index: &ProverIndex<G>,
         prev_challenges: Vec<(Vec<ScalarField<G>>, PolyComm<G>)>,
+        blinders: [Option<PolyComm<ScalarField<G>>>; COLUMNS],
```
Comment (o1-david) [inline kimchi/src/prover.rs:83]: why are you doing this?

Comment (imeckler) [reply kimchi/src/prover.rs:83]: If you want we could make it a separate method with a scary name? It would be nice to merge it in though

Comment (imeckler) [reply kimchi/src/prover.rs:83]: If you want we could make it a separate method with a scary name? My preference would be to merge it but it's OK if we don't

File: groupmap/src/lib.rs [hunk 9]
```diff
@@ -28,6 +28,7 @@ pub trait GroupMap<F> {
     fn batch_to_group_x(&self, ts: Vec<F>) -> Vec<[F; 3]>;
 }
 
+#[derive(Clone, Copy)]
```
[the remark this answers is not in the document]
Comment (mimoo) [reply groupmap/src/lib.rs:31]: agree, but for different PR/issue :o

File: kimchi/src/prover.rs [hunk 10]
```diff
@@ -179,10 +183,11 @@ where
             // padding
             w.extend(std::iter::repeat(ScalarField::<G>::zero()).take(length_padding));
 
+            /*
             // zk-rows
             for row in w.iter_mut().rev().take(ZK_ROWS as usize) {
                 *row = <ScalarField<G> as UniformRand>::rand(rng);
-            }
+            } */
```
Comment (mimoo) [inline kimchi/src/prover.rs:190]: I guess I don't understand why removing these and adding the zk rows manually on the circuit side, is this a question of control as well for the bba thing?

File: circuit-construction/src/lib.rs [hunk 11]
```diff
@@ -0,0 +1,988 @@
...
+        let num_bits = 255;
+        let num_row_pairs = num_bits / 5;
+        let mut witness: [Vec<F>; COLUMNS] = array_init(|_| vec![]);
+
+        let acc0 = self.add_group(zero, (xt, yt), (xt, yt));
+
+        let _ = self.var(|| {
+            witness = array_init(|_| vec![F::zero(); 2 * num_row_pairs]);
+            let bits_msb: Vec<bool> = scalar
+                .0
```
[the remark this answers is not in the document]
Comment (mimoo) [reply circuit-construction/src/lib.rs:459]: resolving these as I'm trying to get this in, we can propose refactors in other PRs :o

File: kimchi/src/prover.rs (hunk 8 above)
Comment (mimoo) [reply kimchi/src/prover.rs]: I added a wrapping `Option<[Option<.>; COLUMNS]>` for now

**Comment by mimoo** [comment]
cc @imeckler can you take a look at this now?

**Review by mimoo — APPROVED** [review]
This looks good to me now, @imeckler can you take a look and cross approve before we merge it?

**Comment by imeckler** [comment]
I cannot approve as I'm the original author, but I will merge :+1:

**Pull request merged**